src/bitset_util.rs

```diff
@@ -14,23 +17,142 @@ const ETHTOOL_A_BITSET_BIT_NAME: u16 = 2;
 ) -> Result<Vec<EthtoolBitSet>, DecodeError> {
-    let error_msg = "failed to parse mode bit sets";
-    for nla in NlasIterator::new(raw) {
-        let nla = &nla.context(error_msg)?;
-        if nla.kind() == ETHTOOL_A_BITSET_BITS {
-            return parse_bitset_bits_nla(nla.value());
+    let is_compact = NlasIterator::new(raw)
+        .filter_map(|nla| nla.ok())
+        .find_map(|nla| match nla.kind() {
+            ETHTOOL_A_BITSET_BITS => Some(false),
+            ETHTOOL_A_BITSET_VALUE | ETHTOOL_A_BITSET_MASK => Some(true),
+            _ => None,
+        })
+        .ok_or("could not determine if bitset is compact")?;
+
+    if is_compact {
+        const BITS_PER_BYTE: usize = 8;
+        let mut size = None;
+        let mut values = Vec::new();
+        let mut masks = Vec::new();
+
+        for nla in NlasIterator::new(raw) {
+            let nla = nla.context("failed to get NLA for compact bitset")?;
+            match nla.kind() {
+                ETHTOOL_A_BITSET_SIZE => {
+                    size = Some(parse_u32(nla.value())? as usize);
+                }
+                ETHTOOL_A_BITSET_VALUE => {
+                    let size =
+                        size.ok_or("compact bitset value: size not set")?;
+                    let bytes = nla.value();
+                    values = (0..size)
+                        .map(|i| {
+                            let byte = bytes
+                                .get(i / BITS_PER_BYTE)
+                                .copied()
+                                .unwrap_or(0);
+                            ((byte >> (i % BITS_PER_BYTE)) & 1) != 0
+                        })
+                        .collect();
+                }
+                ETHTOOL_A_BITSET_MASK => {
+                    let size =
+                        size.ok_or("compact bitset mask: size not set")?;
+                    let bytes = nla.value();
+                    masks = (0..size)
+                        .map(|i| {
+                            let byte = bytes
+                                .get(i / BITS_PER_BYTE)
+                                .copied()
+                                .unwrap_or(0);
+                            ((byte >> (i % BITS_PER_BYTE)) & 1) != 0
+                        })
+                        .collect();
+                }
+                _ => {}
+            }
         }
+
+        let size = size.ok_or("compact bitset: size not set")?;
+        if values.len() != size {
+            return Err("compact bitset: values length mismatch".into());
+        }
+        if masks.len() != size {
+            return Err("compact bitset: masks length mismatch".into());
+        }
+
+        Ok((0..size)
+            .map(|i| {
+                EthtoolBitSet::Compact(EthtoolCompactBitSet {
+                    index: i as u32,
+                    value: values[i],
+                    mask: masks[i],
+                })
+            })
+            .collect())
```

Comment:
Could also be written as

```suggestion
         Ok(values
            .iter()
            .zip(masks)
            .enumerate()
            .map(|(i, (&value, mask))| {
                EthtoolBitSet::Compact(EthtoolCompactBitSet {
                    index: i as u32,
                    value,
                    mask,
                })
            })
            .collect())
```